### PR TITLE
Platform changer

### PR DESCRIPTION
**Pull request by Najsr**
Added feature that can change your platform / client version as described in https://github.com/ReSpeak/TS3Hook/issues/17

### CODE REVIEW

File: TS3Hook/dllmain.cpp [hunk 1]
```diff
@@ -204,6 +223,47 @@ void STD_DECL log_out_packet(char* packet, int length)
...
+		
+		int length_difference = buffer.size() - in_str.size();
+		printf("LENGTH DIFFERENCE: %i\n", length_difference);
+		if (length_difference >= 0) {
+			memcpy(packet, in_str.c_str(), in_str.length());
+			memset(packet + in_str.length(), ' ', length - in_str.length());
+		}
+		else if (nickname_length > 3 && length_difference + nickname_length >= 0) {
+			in_str.erase(client_nickname + 16, (client_ver - client_nickname - 17));
+			in_str.insert(client_nickname + 16, "HAX");
```
Comment (Bluscream) [inline TS3Hook/dllmain.cpp]: Please don't try to edit anything else than `client_version`, `client_platform` and `client_version_sign`

**Review by Bluscream — CHANGES_REQUESTED** [review]

File: TS3Hook/dllmain.cpp [hunk 2]
```diff
@@ -98,9 +100,20 @@ void create_config(const LPCWSTR file_name)
 	WritePrivateProfileString(lpSection, L"ignorecmds", L"", file_name);
 	WritePrivateProfileString(lpSection, L"blockcmds", L"", file_name);
 	WritePrivateProfileString(lpSection, L"injectcmd", L" msg=~cmd", file_name);
+	WritePrivateProfileString(lpSection, L"clientversion", L"", file_name);
```
Comment (Bluscream) [inline TS3Hook/dllmain.cpp:146]: Please add a default value here

**Review by Bluscream — CHANGES_REQUESTED** [review]

File: TS3Hook/dllmain.cpp (hunk 1 above)
Comment (Najsr) [reply TS3Hook/dllmain.cpp]: This makes it possible because it gives it some space which it needs to go on some older versions eg 3.0.19.4

Comment (Bluscream) [reply TS3Hook/dllmain.cpp]: hmm i would accept that when your code doesn't modify the clientinit while the config value is empty

File: TS3Hook/dllmain.cpp (hunk 2 above)
Comment (Bluscream) [reply TS3Hook/dllmain.cpp:146]: Forget it. don't add a default value as i want this to be disabled by default as said in https://github.com/ReSpeak/TS3Hook/pull/18#discussion_r161377982

Just add a comment to the code how the config line should be formatted

`version|platform|versionsign`?

Comment (Najsr) [reply TS3Hook/dllmain.cpp]: Will take a look at this in a moment

File: TS3Hook/dllmain.cpp (hunk 1 above)
Comment (Najsr) [reply TS3Hook/dllmain.cpp]: Done, check last commit

File: TS3Hook/dllmain.cpp (hunk 2 above)
Comment (Najsr) [reply TS3Hook/dllmain.cpp:146]: Dunno how to do that

Comment (Bluscream) [reply TS3Hook/dllmain.cpp:146]: What is your code replacing with what then? just give me a quick list of 
- change = changed

Comment (Najsr) [reply TS3Hook/dllmain.cpp:146]: I don't understand what you want from me :D

Comment (Bluscream) [reply TS3Hook/dllmain.cpp:146]: What does your pull request even do. Thats what i want (to know)

Comment (Najsr) [reply TS3Hook/dllmain.cpp:146]: It adds possibility to change (fake) client platform which can be set in HookConf.ini file

**Comment by Najsr** [comment]
I have added a feature that sets your previous nickname when your nickname was change duo to char limitation. Hope you find it useful :)

**Review by Splamy — CHANGES_REQUESTED** [review]
The problem with adding the normal api is that it might not work when using this as a plugin in lower client versions, the idea behind the the auto api version finder is that you *don't* depend on the api version.
Another problem is that this will crash when using the injector instead of as plugin

**Comment by Bluscream** [comment]
yeah and i don't see where the nickname change would make sense. wouldn't that throw a "invalid parameter size"?

**Comment by Najsr** [comment]
Well if new string is longer than default one (eg. u want 3.0.19.4 version) then your nick is changed to shortest possible one (3 chars) so you can connect without crashing the client (Splamy explained me that the length of the edited string you edit must not exceed length parameter)

**Comment by Najsr** [comment]
Also why would you use injector when you can load it as a plugin?

**Comment by Bluscream** [comment]
Thats a tough question ^^

**Comment by Splamy** [comment]
@Najsr compatibility, when the api detector doesnt work, and it's really convenient for debugging and developing. I can simply run the project and it is injected instead of copying every time

**Comment by Najsr** [comment]
That means it is useful for developers not end users so I think there is no need to worry

**Comment by Bluscream** [comment]
Ts3hook was intended to be for developers only.

**Comment by Najsr** [comment]
If you say so feel free to deny this pull request

**Comment by Bluscream** [comment]
The idea is cool but id prefer the search/replace way from #17. That way platform isnt the only changeable var

**Comment by Splamy** [comment]
A only search/replace isnt very easy either since you can never write a message which is larger then the current packet. I think this pr as it was before was actually fine

**Comment by Najsr** [comment]
@Splamy Did you try to use injector to hook with my newest code? Because it doesn't crash so there is no need to worry. But the auto rename function will only work if you load it as a plugin.

**Comment by Bluscream** [comment]
So the only thing you can currently change is the platform?

```ini
clientversion=Linux
```

?

**Comment by Najsr** [comment]
You can change with which version of teamspeak you want to connect. You just need a name, platform and a sign. Form: Name|Platform|Sign

**Comment by Bluscream** [comment]
What happens if it's empty?

**Comment by Najsr** [comment]
Nothing will happen, it will connect with your installed version

**Comment by Bluscream** [comment]
And what does it do to your nickname when you're using it?

**Comment by Najsr** [comment]
Well if you are changing to a platform that has more characters (eg. 3.0.19.4 >3.1.7) it will try to change your name to 3 chars to make that change possible. And if you install it as a plugin it will change your name back

**Comment by Bluscream** [comment]
Can you change the 3 chars to be random. I don't want this to be detect- and bannable.

**Comment by Bluscream** [comment]
Also i dont understand why injected commands can be so large if the size matters

**Comment by Najsr** [comment]
This is not injecting commands, this is just editing clientinit packet which has given size and you can't exceed it. Also I'll make update tomorrow to randomize nick

**Comment by Bluscream** [comment]
Injected commands are also jut edited sendtextmsg commands

**Comment by Bluscream** [comment]
Thanks and that question was more for @Splamy

**Comment by Najsr** [comment]
It removes part of a message so it will never exceed the length

**Comment by Bluscream** [comment]
When you make that update with randomnick i am fine with merging it, just splamy has to agree

**Review by Bluscream — APPROVED** [review]

**Comment by Bluscream** [comment]
@Splamy please decide if you like it and change it if you don't :)

**Comment by Najsr** [comment]
Sorry for the delay... But I got the randomizer up and running now :)

**Comment by Bluscream** [comment]
Thanks. As it doesn't look like Splamy is gonna make a decision i'm gonna merge it tomorrow.

**Comment by Najsr** [comment]
Would you mind if i added myself into authors list?

**Comment by Bluscream** [comment]
No

**Review by Splamy — APPROVED** [review]

**Comment by Splamy** [comment]
I just wanted to test it before approving, but lgtm.

**Comment by Bluscream** [comment]
Thanks ^^

**Comment by Bluscream** [comment]
@Najsr You might want to increase the version tho :o
@Splamy Can we get a release for this after the version was increased?

**Comment by Splamy** [comment]
@Bluscream done
Btw, why do you always add the Injector.exe alone in your releases. The Injector without the appropriate dll has no use.

**Comment by Bluscream** [comment]
Why should we provide the dll 3 times? But okay, i'll zip them next time

**Comment by Splamy** [comment]
because then you have to download the ts_plugin *seperately*, extract it and move the two files together, how is that any intuitive?

**Comment by Bluscream** [comment]
True

**Comment by Bluscream** [comment]
Just a thought: can't you make the dll a executable at the same time? So one just renames the dll to exe and executes it?

**Comment by Splamy** [comment]
I don't think that's possible, especially since one is C# an the other one C++, and i dont think you can export stuff in exe like with c# where everything is basically the same

**Comment by Bluscream** [comment]
> Btw, why do you always add the Injector.exe alone in your releases. The Injector without the appropriate dll has no use.

On the other hand; why don't you include the .ts3_plugin file in your releases?

**Comment by Splamy** [comment]
No idea how ts3plugins are built correctly, I'll look it up for the next time xD

**Comment by Bluscream** [comment]
Loool

**Comment by Bluscream** [comment]
@Najsr There's one small issue.
When you send a textmessage with just the content `clientinit` TS3Hook does this:
```
[OUT] sendtextmessag3.?.?\s[Build:\s5680278000]
[IN ] error id=256 msg=command\snot\sfound
```

**Comment by Bluscream** [comment]
@Najsr another user (@SossenSystems) pointed out that the platform changer will revert the output_muted status to 0 when connecting so you're no longer muted.

**Pull request merged**